### PR TITLE
Sema: Re-introduce the hack for re-using generic signature of extended protocol

### PR DESCRIPTION
**Pull request by slavapestov**
In Swift 5.10 if you wrote `extension Foo {}` for some protocol Foo,
the extension would always re-use the generic signature of Foo, which
is <Self where Self: Foo>. In Swift 6 this no longer works because Foo
might be ~Copyable, in which case `extension Foo {}` adds default
requirements, so we changed GenericSignatureRequest to just always
build a new signature if we're given an extension.

However, to avoid a request cycle with a code example that really should
have never worked at all, I'm re-introducing the hack for re-using the
signature.

Fixes rdar://problem/129540617.

### CODE REVIEW

File: lib/Sema/TypeCheckGeneric.cpp [hunk 1]
```diff
@@ -821,6 +790,16 @@ GenericSignatureRequest::evaluate(Evaluator &evaluator,
       parentSig = extendedNominal->getGenericSignatureOfContext();
       genericParams = nullptr;
     }
+
+    // The inherited entries influence the generic signature of the extension,
+    // because if it introduces conformance to invertible protocol IP, we do not
+    // we do not infer any requirements that the generic parameters to conform
+    // to invertible protocols. This forces people to write out the conditions.
+    if (ext->hasInvertibleLocalConformance())
+      inferInvertibleReqs = false;
```
Comment (AnthonyLatsis) [inline lib/Sema/TypeCheckGeneric.cpp:822]: > I'm re-introducing the hack for re-using the signature.

Is this not just a nicer way to express what that loop is doing? I.e., we are also cancelling inference of conformance requirements to invertible protocols upon finding an invertible protocol in the inheritance clause there.

Comment (slavapestov) [reply lib/Sema/TypeCheckGeneric.cpp:822]: The actual change here is the last commit.

Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckGeneric.cpp:822]: Oops, sorry, I completely missed that one!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**